src/main/java/org/assertj/core/internal/Iterables.java

```diff
@@ -453,16 +453,62 @@ public void assertContainsOnlyNulls(AssertionInfo info, Iterable<?> actual) {
    * @throws AssertionError if the given {@code Iterable} does not contain the given sequence of objects.
    */
   public void assertContainsSequence(AssertionInfo info, Iterable<?> actual, Object[] sequence) {
-    if (commonCheckThatIterableAssertionSucceeds(info, actual, sequence)) return;
-    // check for elements in values that are missing in actual.
-    List<?> actualAsList = newArrayList(actual);
-    for (int i = 0; i < actualAsList.size(); i++) {
-      // look for given sequence in actual starting from current index (i)
-      if (containsSequenceAtGivenIndex(actualAsList, sequence, i)) return;
+    // perform the checks that would have been done in commonCheckThatIterableAssertionSucceeds
+    // but do them explicitly without having to create a new iterator on actual - which would
+    // break if actual were only singly-traversable.
+    checkNotNullIterables(info, actual, sequence);
+    // store the elements from actual that have been visited (because we don't know we can look ahead - the 'actual'
+    // might be singly-traversable) in a cyclic list, which can then be rolled for each new element
+    // until a match is found or until the 'actual' is exhausted. Of course if 'actual' really is infinite then this could
+    // take a while :-D
+    // once it's full, check if it equals sequence for each new element
+    RingBuffer buffer = new RingBuffer(sequence.length);
+    int howMany=0;
+    final Iterator<?> actual_iterator = actual.iterator();
+    // ... perform the remaining "normal" checks explicitly
+    if(!actual_iterator.hasNext() && sequence.length==0) return;
+    failIfEmptySinceActualIsNotEmpty(sequence);
+    while(actual_iterator.hasNext()) {
+      final Object next = actual_iterator.next();
+      buffer.add(next);
+      howMany++;
+      if(howMany>=sequence.length && buffer.equals(sequence)) return;
     }
     throw actualDoesNotContainSequence(info, actual, sequence);
   }
 
+  class RingBuffer {
+    private int writePosition;
+    private int length;
+    private Object[] buffer;
```

Comment:
have you considered using a linked list? it exposes `addLast` and `removeFirst` methods which can simplify the `add` method and get rid of `writePosition` (I haven't tried it so I might be wrong but worth considering)

Reply:
Hey, interesting idea. I confess that I haven't considered that in detail. I do quite like the RingBuffer because it does exactly what I want to show, nothing more and nothing less. That would certainly not be true of the linked list, which would also need (possibly) the tail pointing to the head. My feeling is that there would be work involved in making the list function as required which would detract from the issue at hand. What do you think?